packages/api/src/controllers/stripe.ts

```diff
@@ -0,0 +1,720 @@
+      billingCycleEnd,
+      ingests[0].origin,
+      adminToken
+    );
+
+    const overUsage = await calculateOverUsage(
+      products[user.stripeProductId],
+      billingUsage
+    );
+
+    const subscriptionItems = await req.stripe.subscriptionItems.list({
+      subscription: user.stripeCustomerSubscriptionId,
+    });
+
+    // create a map of subscription items by their lookup keys
+    const subscriptionItemsByLookupKey = subscriptionItems.data.reduce(
+      (acc, item) => {
+        acc[item.price.lookup_key] = item.id;
+        return acc;
+      },
+      {} as Record<string, string>
+    );
+
+    try {
+      // Invoice items based on overusage
+      await Promise.all(
+        products[user.stripeProductId].usage.map(async (product) => {
+          if (product.name === "Transcoding") {
+            await req.stripe.subscriptionItems.createUsageRecord(
+              subscriptionItemsByLookupKey["transcoding_usage"],
+              {
+                quantity: parseInt(overUsage.TotalUsageMins.toFixed(0)),
+                timestamp: Math.floor(new Date().getTime() / 1000),
+                action: "set",
+              }
+            );
+          } else if (product.name === "Delivery") {
+            await req.stripe.subscriptionItems.createUsageRecord(
+              subscriptionItemsByLookupKey["tstreaming_usage"],
+              {
+                quantity: parseInt(overUsage.DeliveryUsageMins.toFixed(0)),
+                timestamp: Math.floor(new Date().getTime() / 1000),
+                action: "set",
+              }
+            );
+          } else if (product.name === "Storage") {
+            await req.stripe.subscriptionItems.createUsageRecord(
+              subscriptionItemsByLookupKey["tstorage_usage"],
+              {
+                quantity: parseInt(overUsage.StorageUsageMins.toFixed(0)),
+                timestamp: Math.floor(new Date().getTime() / 1000),
+                action: "set",
+              }
+            );
+          }
+        })
+      );
+      updatedUsers.push({
+        id: user.id,
+        email: user.email,
+        overUsage: overUsage,
+        usage: billingUsage,
+        usageReported: true,
+      });
+    } catch (e) {
+      console.log(`
+        Failed to create usage record for user=${user.id} with error=${e.message} - it's pay as you go subscription probably needs to get migrated
+      `);
+      updatedUsers.push({
+        id: user.id,
+        email: user.email,
+        overUsage: overUsage,
+        usage: billingUsage,
+        usageReported: false,
+        error: e.message,
+      });
+    }
+  }
+
+  return {
+    updatedUsers: updatedUsers,
+  };
+};
+
+const calculateOverUsage = async (product, usage) => {
+  const limits = {
+    transcoding: product?.usage[0].limit,
+    streaming: product?.usage[1].limit,
+    storage: product?.usage[2].limit,
+  };
+
+  const overUsage = {
+    TotalUsageMins: Math.max(usage?.TotalUsageMins - limits.transcoding, 0),
```

Comment:
**nit:** Might make sense to remove all the `?` in this func since the objs are never null